# Task 3

First, update the following cell to have a link to your Homework 1 GitHub repository:



**UPDATE THIS LINK:**


https://github.com/annieradams/eds220-hwk-1



In this task, make at least 5 commits at moments you think are adequate. We'll look at your repo to check commit history.

Add comments for all your code. Err on the side of commenting too much for now. 

Comments in each exercise and commits will count towards your grade.

This task covers topics we will see on Monday Oct. 9. The best strategy to solve it is to get together with your classmates and work on it together after Monday's lecture. 


## 1
For this task we are going to use data about  [Western Indian Ocean Coral Diversity (McClanahan, 2023)](https://knb.ecoinformatics.org/view/doi%3A10.5063%2FF1K35S3H) from the Knowledge Network for Biocomplexity (KNB) data repository. 

a. Navigate to the dataset's link and explore the data package. 
 
b. Use the next markdown cell to add a brief description of the dataset, including a citation, date of access, and a link to the archive. 


This dataset includes the following variables for each 6.25 km ^2 reef cell averaged over the 1998-2022 sampling period: coral reef locations, number of coral taxa, climate change/water quality/ human resource management and extraction. The data was collected in Kenya, Tanzania, Mozambique, Madagascar, Comoros, Mayotte, French Souther territories, Mauritius, Reunion, Seychelles, South Africa, Latitude S1.74 to 27.02, and Longitude, E32.89 to 57.8. This data has non sensitive data. This data was accessed on October 10th, 2023. 

Citation:
McClanahan. T.R. 2022. Western Indian Ocean Coral Diversity Data.

Link to archive: 


## 2
In the next code cell, import the pandas package using standard abbreviation in a code cell.

Read in the data from the URL using the `pd.read_csv` function and store it as `coral_div` like this:

```
# read in data
coral_div = pd.read_csv('the URL goes here')
```

In [5]:
# load necessary packages
import pandas as pd

In [6]:
#read in coral data
coral_div = pd.read_csv('https://knb.ecoinformatics.org/knb/d1/mn/v2/object/urn%3Auuid%3Ae7d12a5a-f1a9-4a3e-aaa5-5222f67e799d')

In [7]:
coral_div.head()

PU_ID     PU_Lat    PU_Lon  Number_of_genera   Country  \
0      7  39.076711 -5.490011                35  Tanzania   
1      7  39.076711 -5.490011                22  Tanzania   
2      7  39.076711 -5.490011                25  Tanzania   
3      7  39.076711 -5.490011                26  Tanzania   
4     10  39.054253 -5.512514                27  Tanzania   

                  Ecoregion   Management           Habitat Observer  Depth  \
0  East African Coral Coast  UnfishedLow             Crest      Tim   11.0   
1  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    2.0   
2  East African Coral Coast  UnfishedLow             Slope  Nyawira    4.0   
3  East African Coral Coast  UnfishedLow             Slope  Nyawira    3.0   
4  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    1.5   

   ...  SST_bimodality  SST_Skewness  ClimateStressModel   Netflow  Indegree  \
0  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
1  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
2  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
3  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
4  ...        0.461297     -0.134548            0.629000  0.011721    112.75   

   Outdegree  Retention  min_yr  max_yr  no_replicates  
0      74.75      14.00    2019    2019              1  
1      74.75      14.00    2010    2015              2  
2      74.75      14.00    2015    2015              1  
3      74.75      14.00    2016    2019              2  
4      51.50     183.75    2010    2019              3  

[5 rows x 42 columns]

## 3

Use attributes and methods of the `coral_div` data frame to obtain the following information about it:

a. how many rows and columns does it have?

b. what are the column names?

c. what are the data types of the columns?

d. display the first 8 rows with all the column names. For this one, you will need to udpate a `pandas` option (see discussion section)

e. some inital info you are interested in. 

Use the code cells marked a-e. For each, add a comment explaining what is the ouput.

In [10]:
# a use shape function to find how many rows/columns there are in the dataframe
# returned as (rows, columns)
coral_div.shape

(575, 42)

In [13]:
# b .columns tells us column names
coral_div.columns

Index(['PU_ID', 'PU_Lat', 'PU_Lon', 'Number_of_genera', 'Country', 'Ecoregion',
       'Management', 'Habitat', 'Observer', 'Depth', 'Coralcover',
       'SSTRateofRise', 'PARmax', 'Calcite', 'Dis_oxygen', 'mean.waveNRJ',
       'Salinity_mean', 'andrello_reef_value', 'andrello_nutrients',
       'mean.npp', 'TT_market_hrs', 'TT_pop_hrs', 'cumDHW', 'SST_Kurtosis',
       'SST_Median', 'ChlorA_median', 'andrello_sediments', 'Current_vel_mean',
       'Diff_attenuation', 'Grav_NP', 'Grav_NC', 'PH', 'SST_bimodality',
       'SST_Skewness', 'ClimateStressModel', 'Netflow', 'Indegree',
       'Outdegree', 'Retention', 'min_yr', 'max_yr', 'no_replicates'],
      dtype='object')

In [8]:
# c
#dtypes returns each column name and the type of data that columns holds
coral_div.dtypes

PU_ID                    int64
PU_Lat                 float64
PU_Lon                 float64
Number_of_genera         int64
Country                 object
Ecoregion               object
Management              object
Habitat                 object
Observer                object
Depth                  float64
Coralcover               int64
SSTRateofRise          float64
PARmax                 float64
Calcite                float64
Dis_oxygen             float64
mean.waveNRJ           float64
Salinity_mean          float64
andrello_reef_value      int64
andrello_nutrients     float64
mean.npp               float64
TT_market_hrs          float64
TT_pop_hrs             float64
cumDHW                 float64
SST_Kurtosis           float64
SST_Median             float64
ChlorA_median          float64
andrello_sediments     float64
Current_vel_mean       float64
Diff_attenuation       float64
Grav_NP                float64
Grav_NC                float64
PH                     float64
SST_bimo

In [11]:
# d
#set option display.max.columns allows us to see all column names
#then use head to view first 8 rows of dataframe
pd.set_option("display.max.columns", None)
coral_div.head(8)

PU_ID     PU_Lat    PU_Lon  Number_of_genera   Country  \
0      7  39.076711 -5.490011                35  Tanzania   
1      7  39.076711 -5.490011                22  Tanzania   
2      7  39.076711 -5.490011                25  Tanzania   
3      7  39.076711 -5.490011                26  Tanzania   
4     10  39.054253 -5.512514                27  Tanzania   
5     10  39.054253 -5.512514                29  Tanzania   
6     15  39.054253 -5.535017                19  Tanzania   
7     39  39.009337 -5.647518                28  Tanzania   

                  Ecoregion   Management           Habitat Observer  Depth  \
0  East African Coral Coast  UnfishedLow             Crest      Tim   11.0   
1  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    2.0   
2  East African Coral Coast  UnfishedLow             Slope  Nyawira    4.0   
3  East African Coral Coast  UnfishedLow             Slope  Nyawira    3.0   
4  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    1.5   
5  East African Coral Coast  UnfishedLow  Lagoon/Back reef      Tim    1.5   
6  East African Coral Coast   Restricted             Crest      Tim    8.0   
7  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    3.0   

   Coralcover  SSTRateofRise     PARmax   Calcite  Dis_oxygen  mean.waveNRJ  \
0          45       0.019214  54.747001  0.001075      4.7495      3.714321   
1          45       0.019214  54.747001  0.001075      4.7495      3.714321   
2          45       0.019214  54.747001  0.001075      4.7495      3.714321   
3          45       0.019214  54.747001  0.001075      4.7495      3.714321   
4          48       0.020051  55.395000  0.001747      4.7510      3.270938   
5          48       0.020051  55.395000  0.001747      4.7510      3.270938   
6          41       0.020051  55.395000  0.001747      4.7510      3.270938   
7          49       0.021358  54.935499  0.001744      4.7540      2.857747   

   Salinity_mean  andrello_reef_value  andrello_nutrients     mean.npp  \
0      35.167646                    0            8.388602  1139.112671   
1      35.167646                    0            8.388602  1139.112671   
2      35.167646                    0            8.388602  1139.112671   
3      35.167646                    0            8.388602  1139.112671   
4      35.102688                    0            6.731860  1126.355957   
5      35.102688                    0            6.731860  1126.355957   
6      35.102688                    0            6.731860  1126.355957   
7      35.106329                    0            2.645199  1150.677856   

   TT_market_hrs  TT_pop_hrs     cumDHW  SST_Kurtosis  SST_Median  \
0       1.383333    0.366667  24.133263     -0.803330      27.435   
1       1.383333    0.366667  24.133263     -0.803330      27.435   
2       1.383333    0.366667  24.133263     -0.803330      27.435   
3       1.383333    0.366667  24.133263     -0.803330      27.435   
4       1.383333    0.366667  25.342901     -0.793681      27.455   
5       1.383333    0.366667  25.342901     -0.793681      27.455   
6       1.383333    0.366667  25.342901     -0.793681      27.455   
7       1.783333    0.333333  30.265219     -0.795494      27.525   

   ChlorA_median  andrello_sediments  Current_vel_mean  Diff_attenuation  \
0       0.465182        53540.007810          0.541991           0.09475   
1       0.465182        53540.007810          0.541991           0.09475   
2       0.465182        53540.007810          0.541991           0.09475   
3       0.465182        53540.007810          0.541991           0.09475   
4       0.700521        38477.105470          0.690454           0.11400   
5       0.700521        38477.105470          0.690454           0.11400   
6       0.592692        38477.105470          0.690454           0.11400   
7       0.656533         3224.024414          0.561659           0.10500   

    Grav_NP    Grav_NC      PH  SST_bimodality  SST_Skewness  \


In [12]:
# e
#find out of many unique values there are for each different column
# inital/broad idea at how much our data might vary
coral_div.nunique()

PU_ID                  304
PU_Lat                 152
PU_Lon                 217
Number_of_genera        37
Country                 11
Ecoregion                6
Management               4
Habitat                  6
Observer                16
Depth                  111
Coralcover              46
SSTRateofRise          283
PARmax                 221
Calcite                219
Dis_oxygen             187
mean.waveNRJ           237
Salinity_mean          226
andrello_reef_value    139
andrello_nutrients     165
mean.npp               197
TT_market_hrs          129
TT_pop_hrs              51
cumDHW                 283
SST_Kurtosis           283
SST_Median             193
ChlorA_median          283
andrello_sediments     180
Current_vel_mean       224
Diff_attenuation       165
Grav_NP                168
Grav_NC                169
PH                      62
SST_bimodality         282
SST_Skewness           283
ClimateStressModel     297
Netflow                126
Indegree               110
O

## 4

Some of the columns have data type `int64` and some have data type `float64`. What is the difference between these `int64` and `float64`?


The difference between int64 and float64 is that int64 is of type integer, meaning that values are whole numbers. Float 64 is of type float, wehre values have fractional values associated with them. The 64 in both these types represents that number of bits it occupies. 

## 5

The `Country` column indicates the country of reef location. Explain in your words what is the ouput of `coral_div.Country` and `coral_div.Country.unique()`. 

coral_div.Country selects the "Country" column and prints that column and the index for the whole dataset. Therefore, we see what the country value is for all 575 rows of the data frame. coral_div.Country.unique() is different in that it it only returns the names of countries that exist in the dataset once, rather than the country value for each individual row. 

## 6

Check the dataset's [metadata file README_md.doc](https://knb.ecoinformatics.org/view/doi:10.5063/F1K35S3H) to find which column has information about who made an observation. How many observers collected this data? HINT: use the `len` Python function. Not sure what `len` does? Use `?`.

In [19]:
#observer column is name of data observer
#find number of unique observers to find out how many observers collected this data
coral_div.Observer.nunique

<bound method IndexOpsMixin.nunique of 0          Tim
1      Nyawira
2      Nyawira
3      Nyawira
4      Nyawira
        ...   
570       Sink
571     Carlos
572     Carlos
573     Porter
574       Sink
Name: Observer, Length: 575, dtype: object>

## 7

The `Coralcover` column indicates the percent of hard coral cover in each 6km reef cell surveyed. What data is `coral_div[coral_div['Coralcover'] >= 50]` selecting?

coral_div[coral_div['Coralcover'] >= 50] is selecting rows in the dataset in which the percentage of coral covered is greater than or equal to 50. 

## 8

Use code to find if there are there any sites with a 30% coral cover in Tanzania.

In [18]:
#select coral cover column, and speciffy that we want values that are qual to exactly 30
# select coral colver column 
coral_div[coral_div['Coralcover'] == 30]["Coralcover"]

63     30
64     30
122    30
123    30
124    30
125    30
126    30
127    30
139    30
210    30
341    30
342    30
343    30
344    30
377    30
388    30
389    30
Name: Coralcover, dtype: int64

## 9

We are interested in looking at the PH of sea water (`PH` column),  mean salinity (PSS) (`Salinity_mean` column), and depth of survey location in meters (`Depth` column). Select this data. 

In [28]:
#select using a list of all column names we want
coral_div[['PH','Salinity_mean','Depth']]

PH  Salinity_mean  Depth
0    8.2175      35.167646  11.00
1    8.2175      35.167646   2.00
2    8.2175      35.167646   4.00
3    8.2175      35.167646   3.00
4    8.2180      35.102688   1.50
..      ...            ...    ...
570  8.2400      35.301208  15.50
571  8.2400      35.301208  12.00
572  8.2400      35.301208  12.60
573  8.2400      35.301208  13.00
574  8.2400      35.301208  13.47

[575 rows x 3 columns]

## 10
The `max_yr` column indicates the year of last sampling at a given site. Select the PH, salinity, and depth for sites where sampling ended in 2017. 


In [42]:
# select values where max year is equal to 2017 ( the year of the last sampling at a site)
#select Ph salinity mean, and depth by creating a list of column names we want returned
coral_div.loc[coral_div['max_yr'] == 2017][['PH','Salinity_mean','Depth']]

PH  Salinity_mean      Depth
67   8.2150      35.100085  16.000000
68   8.2145      35.144927  10.000000
70   8.2150      35.101086  11.000000
71   8.2150      35.101086  11.000000
72   8.2150      35.101086   9.000000
73   8.2150      35.101086   9.000000
84   8.2110      35.278084  12.500000
85   8.2110      35.278084   6.000000
164  8.2220      35.337914   9.875000
167  8.2220      35.337914  12.075000
169  8.2220      35.337914  10.300000
172  8.2220      35.337914  11.000000
174  8.2225      35.287419   7.266667
176  8.2225      35.287419  13.225000
179  8.2225      35.287419  10.550000
182  8.2220      35.331669  12.033333
185  8.2230      35.374460  10.900000
188  8.2230      35.374460  14.725000
191  8.2230      35.247482   9.725000
194  8.2230      35.341358  11.075000
197  8.2230      35.390354  10.600000
346  8.2165      35.114244   7.000000
347  8.2165      35.112398   3.000000
348  8.2165      35.112398   4.000000
349  8.2160      35.117374   5.500000
350  8.2160      35.117374   2.000000
351  8.2160      35.117374  15.000000
352  8.2160      35.117374   5.000000
353  8.2165      35.149462  11.000000
354  8.2160      35.270199   9.000000
395  8.2400      35.276642   9.416667
559  8.2120      35.204185   1.500000
566  8.2400      35.301208  17.000000
569  8.2400      35.301208  13.833333
574  8.2400      35.301208  13.470000

## 11
Use `iloc` to select rows 95 throuhg 100 (including 100) in the 17th, 20th, and 28th columns and assign these values to a new variable called `subset`. HINT: the ouput columns should be `Salinity_mean`, `mean.npp` and `Current_vel_mean`. Are you indexing from 0?

In [21]:
#.iloc[rows, columns], use list when selecting multiple rows/columns
subset = coral_div.iloc[[95, 96, 97, 98, 99, 100],[16,19,27]]
subset

Salinity_mean     mean.npp  Current_vel_mean
95       35.201630  1365.146729          0.878417
96       35.204185  1117.635010          0.839372
97       35.234774  1105.270142          0.685955
98       35.234774  1105.270142          0.685955
99       35.250068  1082.196411               NaN
100      35.250068  1082.196411               NaN

## 12 

Run `subset.isna()`. Explain what this code does and what the output represents.

In [22]:
#use is.na() to find out if each individual cell has an na value or not
#true if value is NA, false if value is not NA
subset.isna()

Salinity_mean  mean.npp  Current_vel_mean
95           False     False             False
96           False     False             False
97           False     False             False
98           False     False             False
99           False     False              True
100          False     False              True

This code tells whether each individual cell has NA value or not. If the cell is populated by "False", this means that there is not a NA in that cell and it instead has a value. If the cell is populated by "true", this means that that observation has a NA value. For example, the 100th observation of Current_vel_mean says "True", meaning that there is a missing value in that particular cell.  


## 13
Run `subset.isna().any()`. Explain what this code does and what the output represents. HINT: It could be helpful to check the [documentation of `pandas.Sries.any`](https://pandas.pydata.org/docs/reference/api/pandas.Series.any.html).


In [23]:
#.any() tells us if there is at least one NA in any given column
subset.isna().any()

Salinity_mean       False
mean.npp            False
Current_vel_mean     True
dtype: bool

This function tells us if there are any missing values for each column in the dataframe. For example, the salinity_mean column and mean.npp column in our subset data frame do not have any missing values. The current_vel_mean column, however, as **at least** one missing value. 

## 14 

Select the rows in `subset` that have NA in the `Current_vel_mean` column. HINT: you can use your answer for 9.

In [25]:
#check for na values, use axis = 1 to get columns
subset[subset.Current_vel_mean.isna()== True]



Salinity_mean     mean.npp  Current_vel_mean
99       35.250068  1082.196411               NaN
100      35.250068  1082.196411               NaN